### PR TITLE
Stop sample previews on mouse release or browser focus loss

### LINKED ISSUE #5736: Crash when previewing sample
Got a crash when previewing a sample when running lmms in gdb.
`-DCMAKE_BUILD_TYPE=Debug -DWANT_DEBUG_FPE=True`
lmms version - master, 7808e0b92fdb7f84180553fa534d66b2f2f041c6 with one unrelated commit on top.

#### Method to replicate
 * Start master, debug build in gdb.
 * Open  demo project loop.mmp [loop.mmp.zip](https://github.com/LMMS/lmms/files/5440854/loop.mmp.zip)
 * Press play in the Song Editor
 * Demo a sample from the side menu
 * Press paus in the Song Editor
 * Demo a sample from the side menu
 * Crash
#### Logs
<details>
  <summary>Click to expand</summary>
<pre>
Thread 1 "lmms" received signal SIGSEGV, Segmentation fault.
0x00005555557b305f in Mixer::removePlayHandle (this=0x555555c920c0, _ph=0x55555a4d34f0) at /home/zonkmachine/builds/lmms/src/core/Mixer.cpp:705
705		if( _ph->affinityMatters() &&
(gdb) bt
#0  0x00005555557b305f in Mixer::removePlayHandle(PlayHandle*) (this=0x555555c920c0, _ph=0x55555a4d34f0)
    at /home/zonkmachine/builds/lmms/src/core/Mixer.cpp:705
#1  0x000055555586a7f9 in FileBrowserTreeWidget::stopPreview() (this=0x5555598e4650) at /home/zonkmachine/builds/lmms/src/gui/FileBrowser.cpp:607
#2  0x000055555586a13b in FileBrowserTreeWidget::previewFileItem(FileItem*) (this=0x5555598e4650, file=0x555559911170)
    at /home/zonkmachine/builds/lmms/src/gui/FileBrowser.cpp:543
#3  0x000055555586a0c2 in FileBrowserTreeWidget::mousePressEvent(QMouseEvent*) (this=0x5555598e4650, me=0x7fffffffd3e0)
    at /home/zonkmachine/builds/lmms/src/gui/FileBrowser.cpp:532
#4  0x00007ffff7a972d1 in QWidget::event(QEvent*) (this=this@entry=0x5555598e4650, event=event@entry=0x7fffffffd3e0) at kernel/qwidget.cpp:8959
#5  0x00007ffff7b44d52 in QFrame::event(QEvent*) (this=0x5555598e4650, e=0x7fffffffd3e0) at widgets/qframe.cpp:550
#6  0x00007ffff7cc9482 in QAbstractItemView::viewportEvent(QEvent*) (this=this@entry=0x5555598e4650, event=event@entry=0x7fffffffd3e0)
    at itemviews/qabstractitemview.cpp:1750
#7  0x00007ffff7d3762f in QTreeView::viewportEvent(QEvent*) (this=0x5555598e4650, event=0x7fffffffd3e0) at itemviews/qtreeview.cpp:1318
#8  0x00007ffff66e664b in QCoreApplicationPrivate::sendThroughObjectEventFilters(QObject*, QEvent*) (event=<optimized out>, receiver=<optimized out>)
    at kernel/qcoreapplication.cpp:1214
#9  QCoreApplicationPrivate::sendThroughObjectEventFilters(QObject*, QEvent*) (receiver=receiver@entry=0x5555598f3850, event=event@entry=0x7fffffffd3e0)
    at kernel/qcoreapplication.cpp:1203
#10 0x00007ffff7a54a55 in QApplicationPrivate::notify_helper(QObject*, QEvent*)
    (this=this@entry=0x555555b12f70, receiver=receiver@entry=0x5555598f3850, e=e@entry=0x7fffffffd3e0) at kernel/qapplication.cpp:3694
#11 0x00007ffff7a5e343 in QApplication::notify(QObject*, QEvent*) (this=<optimized out>, receiver=0x5555598f3850, e=0x7fffffffd3e0)
    at kernel/qapplication.cpp:3160
#12 0x00007ffff66e693a in QCoreApplication::notifyInternal2(QObject*, QEvent*) (receiver=0x5555598f3850, event=0x7fffffffd3e0)
    at ../../include/QtCore/../../src/corelib/kernel/qobject.h:142
#13 0x00007ffff7a5d457 in QApplicationPrivate::sendMouseEvent(QWidget*, QMouseEvent*, QWidget*, QWidget*, QWidget**, QPointer<QWidget>&, bool, bool)
    (receiver=receiver@entry=0x5555598f3850, event=event@entry=0x7fffffffd3e0, alienWidget=alienWidget@entry=0x5555598f3850, nativeWidget=0x555559501330, buttonDown=buttonDown@entry=0x7ffff7f848d0 <qt_button_down>, lastMouseReceiver=..., spontaneous=true, onlyDispatchEnterLeave=false)
    at kernel/qapplication.cpp:2646
#14 0x00007ffff7ab335d in QWidgetWindow::handleMouseEvent(QMouseEvent*) (this=0x555559e13700, event=0x7fffffffd860)
    at /usr/include/c++/9/bits/atomic_base.h:413
#15 0x00007ffff7ab61ec in QWidgetWindow::event(QEvent*) (event=0x7fffffffd860, this=0x555559e13700) at kernel/qwidgetwindow.cpp:289
#16 QWidgetWindow::event(QEvent*) (this=0x555559e13700, event=0x7fffffffd860) at kernel/qwidgetwindow.cpp:232
#17 0x00007ffff7a54a66 in QApplicationPrivate::notify_helper(QObject*, QEvent*)
    (this=this@entry=0x555555b12f70, receiver=receiver@entry=0x555559e13700, e=e@entry=0x7fffffffd860) at kernel/qapplication.cpp:3700
#18 0x00007ffff7a5e0f0 in QApplication::notify(QObject*, QEvent*) (this=0x555555b12f50, receiver=0x555559e13700, e=0x7fffffffd860)
    at kernel/qapplication.cpp:3446
#19 0x00007ffff66e693a in QCoreApplication::notifyInternal2(QObject*, QEvent*) (receiver=0x555559e13700, event=0x7fffffffd860)
    at ../../include/QtCore/../../src/corelib/kernel/qobject.h:142
#20 0x00007ffff6acf7d3 in QGuiApplicationPrivate::processMouseEvent(QWindowSystemInterfacePrivate::MouseEvent*) (e=e@entry=0x55555a818520)
    at kernel/qguiapplication.cpp:2107
#21 0x00007ffff6ad110b in QGuiApplicationPrivate::processWindowSystemEvent(QWindowSystemInterfacePrivate::WindowSystemEvent*) (e=e@entry=0x55555a818520)
    at kernel/qguiapplication.cpp:1842
#22 0x00007ffff6aab35b in QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) (flags=flags@entry=...)
    at kernel/qwindowsysteminterface.cpp:1151
#23 0x00007ffff214032e in xcbSourceDispatch(GSource*, GSourceFunc, gpointer) (source=<optimized out>) at qxcbeventdispatcher.cpp:105
#24 0x00007ffff5455fbd in g_main_context_dispatch () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#25 0x00007ffff5456240 in  () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#26 0x00007ffff54562e3 in g_main_context_iteration () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#27 0x00007ffff673e565 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) (this=0x555555bcd490, flags=...)
    at kernel/qeventdispatcher_glib.cpp:422
#28 0x00007ffff66e54db in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) (this=this@entry=0x7fffffffdc00, flags=..., flags@entry=...)
    at ../../include/QtCore/../../src/corelib/global/qflags.h:140
#29 0x00007ffff66ed246 in QCoreApplication::exec() () at ../../include/QtCore/../../src/corelib/global/qflags.h:120
#30 0x000055555571bbd4 in main(int, char**) (argc=1, argv=0x7fffffffdf78) at /home/zonkmachine/builds/lmms/src/core/main.cpp:991
(gdb) bt full
#0  0x00005555557b305f in Mixer::removePlayHandle(PlayHandle*) (this=0x555555c920c0, _ph=0x55555a4d34f0)
    at /home/zonkmachine/builds/lmms/src/core/Mixer.cpp:705
#1  0x000055555586a7f9 in FileBrowserTreeWidget::stopPreview() (this=0x5555598e4650) at /home/zonkmachine/builds/lmms/src/gui/FileBrowser.cpp:607
        previewLocker = {val = 93825063077529}
#2  0x000055555586a13b in FileBrowserTreeWidget::previewFileItem(FileItem*) (this=0x5555598e4650, file=0x555559911170)
    at /home/zonkmachine/builds/lmms/src/gui/FileBrowser.cpp:543
        previewLocker = {val = 93825063077529}
        newPPH = 0x5555598f1a30
        fileName = {static null = {<No data fields>}, d = 0x7fffffffcc40}
        ext = {static null = {<No data fields>}, d = 0x555559911170}
#3  0x000055555586a0c2 in FileBrowserTreeWidget::mousePressEvent(QMouseEvent*) (this=0x5555598e4650, me=0x7fffffffd3e0)
    at /home/zonkmachine/builds/lmms/src/gui/FileBrowser.cpp:532
        i = 0x555559911170
        f = 0x555559911170
#4  0x00007ffff7a972d1 in QWidget::event(QEvent*) (this=this@entry=0x5555598e4650, event=event@entry=0x7fffffffd3e0) at kernel/qwidget.cpp:8959
        d = 0x5555598f1a30
#5  0x00007ffff7b44d52 in QFrame::event(QEvent*) (this=0x5555598e4650, e=0x7fffffffd3e0) at widgets/qframe.cpp:550
        result = <optimized out>
#6  0x00007ffff7cc9482 in QAbstractItemView::viewportEvent(QEvent*) (this=this@entry=0x5555598e4650, event=event@entry=0x7fffffffd3e0)
    at itemviews/qabstractitemview.cpp:1750
        d = <optimized out>
#7  0x00007ffff7d3762f in QTreeView::viewportEvent(QEvent*) (this=0x5555598e4650, event=0x7fffffffd3e0) at itemviews/qtreeview.cpp:1318
        d = <optimized out>
#8  0x00007ffff66e664b in QCoreApplicationPrivate::sendThroughObjectEventFilters(QObject*, QEvent*) (event=<optimized out>, receiver=<optimized out>)
    at kernel/qcoreapplication.cpp:1214
        obj = <optimized out>
        i = 0
#9  QCoreApplicationPrivate::sendThroughObjectEventFilters(QObject*, QEvent*) (receiver=receiver@entry=0x5555598f3850, event=event@entry=0x7fffffffd3e0)
    at kernel/qcoreapplication.cpp:1203
#10 0x00007ffff7a54a55 in QApplicationPrivate::notify_helper(QObject*, QEvent*)
    (this=this@entry=0x555555b12f70, receiver=receiver@entry=0x5555598f3850, e=e@entry=0x7fffffffd3e0) at kernel/qapplication.cpp:3694
        consumed = false
        filtered = false
#11 0x00007ffff7a5e343 in QApplication::notify(QObject*, QEvent*) (this=<optimized out>, receiver=0x5555598f3850, e=0x7fffffffd3e0)
    at kernel/qapplication.cpp:3160
        me = 
              {<QInputEvent> = {<QEvent> = {_vptr.QEvent = 0x7ffff6f7d200 <vtable for QMouseEvent+16>, static staticMetaObject = {d = {superdata = 0x0, stringdata = 0x7ffff688cde0 <qt_meta_stringdata_QEvent>, data = 0x7ffff688c820 <qt_meta_data_QEvent>, static_metacall = 0x0, relatedMetaObjects = 0x0, extradata = 0x0}}, d = 0x0, t = 2, posted = 0, spont = 1, m_accept = 1, reserved = 2876}, modState = {i = 0}, ts = 5706568}, l = {xp = 136, yp = 567}, w = {xp = 172, yp = 728}, s = {xp = 172, yp = 751}, b = Qt::LeftButton, mouseState = {i = 1}, caps = 0, velocity = {v = {0, 0}}}
        w = 0x5555598f3850
        mouse = 0x7fffffffd3e0
        eventAccepted = <optimized out>
        relpos = {xp = 136, yp = 567}
        pw = {wp = {d = 0x55555a48a690, value = 0x5555598f3850}}
        d = <optimized out>
        res = false
#12 0x00007ffff66e693a in QCoreApplication::notifyInternal2(QObject*, QEvent*) (receiver=0x5555598f3850, event=0x7fffffffd3e0)
    at ../../include/QtCore/../../src/corelib/kernel/qobject.h:142
        selfRequired = true
        result = false
        cbdata = {0x5555598f3850, 0x7fffffffd3e0, 0x7fffffffd1df}
--Type <RET> for more, q to quit, c to continue without paging--c
        d = <optimized out>
        threadData = 0x555555b130f0
        scopeLevelCounter = {threadData = 0x555555b130f0}
#13 0x00007ffff7a5d457 in QApplicationPrivate::sendMouseEvent(QWidget*, QMouseEvent*, QWidget*, QWidget*, QWidget**, QPointer<QWidget>&, bool, bool) (receiver=receiver@entry=0x5555598f3850, event=event@entry=0x7fffffffd3e0, alienWidget=alienWidget@entry=0x5555598f3850, nativeWidget=0x555559501330, buttonDown=buttonDown@entry=0x7ffff7f848d0 <qt_button_down>, lastMouseReceiver=..., spontaneous=true, onlyDispatchEnterLeave=false) at kernel/qapplication.cpp:2646
        graphicsWidget = false
        widgetUnderMouse = <optimized out>
        wasLeaveAfterRelease = true
        result = true
#14 0x00007ffff7ab335d in QWidgetWindow::handleMouseEvent(QMouseEvent*) (this=0x555559e13700, event=0x7fffffffd860) at /usr/include/c++/9/bits/atomic_base.h:413
        translated = {<QInputEvent> = {<QEvent> = {_vptr.QEvent = 0x7ffff6f7d200 <vtable for QMouseEvent+16>, static staticMetaObject = {d = {superdata = 0x0, stringdata = 0x7ffff688cde0 <qt_meta_stringdata_QEvent>, data = 0x7ffff688c820 <qt_meta_data_QEvent>, static_metacall = 0x0, relatedMetaObjects = 0x0, extradata = 0x0}}, d = 0x0, t = 2, posted = 0, spont = 1, m_accept = 1, reserved = 2870}, modState = {i = 0}, ts = 5706568}, l = {xp = 136, yp = 567}, w = {xp = 172, yp = 728}, s = {xp = 172, yp = 751}, b = Qt::LeftButton, mouseState = {i = 1}, caps = 0, velocity = {v = {0, 0}}}
        contextMenuTrigger = QEvent::MouseButtonPress
        widget = 0x5555598f3850
        mapped = {xp = 136, yp = 567}
        initialPress = <optimized out>
        receiver = 0x5555598f3850

</pre>
</details>

_Edit: method to replicate issue updated_

### PR DESCRIPTION
**Pull request by Spekular**
This fix/workaround closes #5736.

I'll make a PR once this is merged that allows samples to keep playing after mouse release (depending on the users settings), because it seems that stopping samples on focus loss is enough to prevent the crash*. In other words, all that remains is some work to add/save/read the option. However, since this would increase the scope of the PR and review times, I'd like to get this merged first so the crash is fixed and nightly builds can go up.

*If there turns out to be a race condition I think we can either add a stopPreview call to the stop button code or fix it in the mixer.

### CODE REVIEW

File: src/gui/FileBrowser.cpp [hunk 1]
```diff
@@ -685,6 +685,11 @@ void FileBrowserTreeWidget::mouseReleaseEvent(QMouseEvent * me )
 			if (s && s->totalFrames() - s->framesDone() <= second * 3)
 			{
 				s->setDoneMayReturnTrue(true);
+				// TODO: This means we can't interrupt this preview, so if a new one 
+				// starts before this one ends they will overlap. Removing this line
+				// leads to a crash (see https://github.com/LMMS/lmms/issues/5736)
+				// right now, but with some fixes to Mixer it should be safe to do.
+				m_previewPlayHandle = NULL;
```
Comment (IanCaio) [inline src/gui/FileBrowser.cpp:692]: Small suggestion

```suggestion
				m_previewPlayHandle = nullptr;
```

**Comment by superpaik** [comment]
Though it fixes the crash, it introduces a somehow annoying side effect (I think it's been explained in the #5736 thread): the loop continues to sound while the new one is playing, which is very confusing, and can be very annoying if the loop is long.
This side-effect doesn't happen when you move around samples with the keyboard, samples go on and off perfectly. Maybe looking at the code for how keyboard is managed can help.

**Comment by Spekular** [comment]
@superpaik The fact that samples (shorter than 3 seconds) continue after mouse release is intentional behavior, and would be very easy to remove. It does make some sense for samples like drums that they can be played just by clicking (rather than holding), but it can be unintuitive so IMO it might be a good idea to remove it.

Another potential fix: Stop previews when the sidebar loses focus (Ableton seems to play entire samples on click and stop them when focus is lost or a new one starts). Theoretically, if the preview is stopped by the sample browser before the stop button is pressed, the issue shouldn't occur. However, depending on the order in which the events are sent and processed the mixer clear function might end up being called before the sidebar calls stopPreview, so this feels like a risky fix (potential race condition).

**Comment by Spekular** [comment]
Or maybe the function hooked up to the stop button could tell the file browser to stop the preview before stopping the song. That should guarantee the order of the calls and prevent the crash.

**Comment by superpaik** [comment]
@Spekular I understand that the sample continue playing after mouserelease, it's a good and desired feature. But I think that when another sample is clicked, the one that is playing should stop (as you said others DAW do).
In fact, that is what happen when you move around samples with keyboard and not mouse (I believe this is something new -keyboard- in the master not available in the stable version (1.2.2).
I'm not very good at coding but on the keyPressEvent function, the first thing that is done is stop another samples from playing. I think that's what is needed inside mousePressEvent. I may be wrong, though.
imatge

**Comment by Spekular** [comment]
I've changed the method used to prevent the crash, so this PR now stops samples on mouse release. **Note**, I intend to add an option for those who prefer the sample to continue playing as soon as this is merged. In other words I don't think this PR needs a long debate about the preferred behavior, just a review of code quality and some testing to ensure that it works as described.

**Comment by Spekular** [comment]
Tested to fix the crash by myself and zonkmachine, and the code is rather simple, so I'll merge this.

**Pull request merged**